ddtrace/tracer.py

```diff
@@ -452,3 +458,70 @@ def set_tags(self, tags):
         :param dict tags: dict of tags to set at tracer level
         """
         self.tags.update(tags)
+
+    def on(self, hook_name):
+        """
+        Decorator used to configure a tracer span hook
+
+        Usage::
+
+            # Register a hook for any emitted `my.span` Spans
+            @tracer.on('my.span')
+            def on_falcon_request(span, data):
+                span.set_tag('key', data['key'])
+
+
+            # Create and emit a `my.span` Span with the provided data
+            with tracer.trace('my.span') as span:
+                tracer.emit(span, dict(key='value'))
+
+        :param hook_name: The name of the hook to configure a hook for
+        :type hook_name: str
+        :rtype: function
+        :returns: A function decorator that registers a function for the provided hook name
+        """
+        def wrapper(func):
+            self.hooks[hook_name].add(func)
+            return func
+        return wrapper
+
+    def emit(self, span, *args, **kwargs):
```

Comment:
I guess this API is used only by our instrumentation right? I don't see the benefit of using it from user code (or maybe yes?). My question is more about if we want to make it "public" or if we want to keep these primitives internal so that only `tracer.on()` is explained and documented.

Reply:
Yeah, I can't really imagine why a user would use this.

I'll move to `def _emit()`, can always make it public in the future if we need to, easier than moving it to being internal.